### DISCUSSION TITLE
Encryption Best Practices

### DISCUSSION THREAD

**Discussion by jacobsomer** [supabase/supabase#9868] [Questions] [answered]
Hello Supabase Team,

Right now, I am building a Supabase service that stores private user information (email, phone, ...). Would the service be more secure if I implement an additional layer of encryption prior to inserting into a Supabase database? An example of encryption could be implementing a one-way hash of passwords on the server before storing the it in the Supabase instance. Is that step redundant? or does your team recommend that I implement in-house encryption on top of Supabase? Thank you for helping out and I look forward to deploying on your platform!

Jacob

**Comment by michelp** [comment, marked as answer]
Hi @jacobsomer ,

For most cases, storing private user information encrypted "at rest" is sufficient for basic data protection, and for postgres in general this means making sure that data written to tables is encrypted, that any backups or dumps contain only encrypted data, and that outgoing replication streams contain only encrypted data.

A very important detail with encryption at rest is that the encrypted stored data *cannot also store the keys* because that would negate the whole purpose.

Good news is we are working on a new product offering that we pre-released a few weeks ago called the [Vault](https://michelp.github.io/pgsodium/Transparent_Column_Encryption.html).  The Vault uses pgsodium's [Transparent Column Encryption](https://michelp.github.io/pgsodium/Transparent_Column_Encryption.html) that automatically creates an encryption trigger and decryption view on columns that you "label" using the standard Postgres [SECURITY LABEL](https://www.postgresql.org/docs/current/sql-security-label.html) syntax.  See the TCE link above for examples.  Because of pgsodium's [Server Key Management](https://github.com/michelp/pgsodium#server-key-management) only key ids are stored next to encrypted data, never the keys themselves.  There's more details on how that works in the link above and in the pgsodium documentation.

The Vault is a simple table named `vault.secrets` that uses TCE to encrypt a secret column, but you can use the same TCE functionality to encrypt one or more of your own (text or bytea) columns on your tables using the same declarative label syntax.  Because [Authenticated Encryption with Associated Data](https://en.wikipedia.org/wiki/Authenticated_encryption#Authenticated_encryption_with_associated_data_(AEAD)) is used, you can also "sign" other non-encrypted columns in a row into the secret column's authentication signature.

So you mentioned hashing, yes you can take these approaches, but it means doing key management and all that stuff in your application layer, you may be more comfortable with that, and you may not want (or be able to) "trust" your database with that data in which case you'd have to do it somewhere upstream, but to me it sounds like the Vault will be a good fit if you're looking for basic data protection "at rest" on disk and in backups.

We are actively working on the Vault as I type, and we intend to have it available for general release at our next launch week, but I'm available for any questions you might have about your specific use cases and we can at least get them on the board for future support.

**Reply by rikur** [reply]
Thanks for the answer! Some of the links are return 404 as they point to michelp.github.io.

When it comes to best practices, a generic guide on how TCE can be used with Supabase Auth for per-user encryption key + nonce could be helpful.

Thanks again :)
